lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -1375,6 +1492,205 @@ makeChangeForCoin
+--      number of large change bundles. By adding to the largest, we are
+--      only increasing the "usefulness" of that bundle.
+--
+-- The solution for burning maintains the same properties:
+--   1. The number of change bundles is not changed, in the case we burn a
+--      change bundle completely, we just leave it as an empty entry
+--      (effectively "pad with zeros").
+--   2. By removing from the smallest bundle, we maintain the ascending partial
+--      ordering of the change bundles.
+--   3. By removing from the smallest bundles, we remove the "least useful"
+--      bundles, maximizing the overall usefulness of our bundles.
+
+-- | Adds a minted value to the change map. This function presumes that the
+-- change map is partially-ordered in ascending order (i.e. the largest
+-- element is at the end of the list).
+--
+-- Properties:
+-- 1. wrt to the total value of assets in the changeMaps, calling this
+-- function is the same as adding the mint value from the change maps:
+--    F.fold changeMaps <> TokenMap.singleton assetId qty
+--          == F.fold (f (assetId, qty) changeMaps)
+-- 2. length is conserved
+-- 3. order is preserved (i.e., the result is in ascending partial order).
+addMintValueToChangeMaps
+    :: (AssetId, TokenQuantity)
+    -> NonEmpty TokenMap
+    -> NonEmpty TokenMap
+addMintValueToChangeMaps (assetId, assetQty) = \case
+    -- This is the largest element, add mint value here
+    (x :| []) ->
+        TokenMap.add (TokenMap.singleton assetId assetQty) x :| []
+    -- There is a larget element, recurse to end of list
+    (x :| (y : rest)) ->
+        let
+            xs = y :| rest
+        in
+          x `NE.cons` addMintValueToChangeMaps (assetId, assetQty) xs
+
+addMintValuesToChangeMaps
+  :: TokenMap
+  -> NonEmpty TokenMap
+  -> NonEmpty TokenMap
+addMintValuesToChangeMaps mints =
+    F.foldr
+        (\mint -> (addMintValueToChangeMaps mint .))
+        id
+        (TokenMap.toFlatList mints)
+
+-- | Remove a burnt value from the change map. This function presumes that
+-- the change map is partially-ordered in ascending order (i.e. the largest
+-- element is at the end of the list). This function operates under the
+-- precondition that a valid selection has been performed, and so assumes
+-- that there is enough asset quantity in the change map to satisfy the
+-- burn. If there is not, it will burn as much as it can.
+--
+-- Properties:
+-- 1. wrt to the total value of assets in the changeMaps, calling this
+-- function is the same as removing the burnt value from the change maps:
+--   F.fold changeMaps `TokenMap.difference` TokenMap.singleton assetId qty
+--          == F.fold (f (assetId, qty) changeMaps)
+-- 2. length is conserved
+-- 3. order is preserved (i.e., the result is in ascending partial order).
+removeBurnValueFromChangeMaps
+    :: (AssetId, TokenQuantity)
+    -> NonEmpty TokenMap
+    -> NonEmpty TokenMap
+removeBurnValueFromChangeMaps (assetId, assetQty) =
```

Comment:
Suggestion: consider simplifying this function with the use of `TokenMap.adjustQuantity`:

```hs
removeBurnValueFromChangeMaps (assetId, assetQty) maps =
    NE.fromList $ burn assetQty (NE.toList maps) []
  where
    burn :: TokenQuantity -> [TokenMap] -> [TokenMap] -> [TokenMap]
    burn _ [] alreadyProcessed = reverse alreadyProcessed
    burn toBurn (this : rest) alreadyProcessed
        | TokenMap.getQuantity this assetId >= toBurn =
            reverse alreadyProcessed <> (this' : rest)
        | otherwise =
            burn toBurn' rest (this' : alreadyProcessed)
      where
        this' =
            TokenMap.adjustQuantity this assetId
            (`TokenQuantity.difference` toBurn)
        toBurn' = TokenMap.getQuantity this assetId
            `TokenQuantity.distance` toBurn
```

Reply:
It's also possible to split up these functions like so:
```hs
removeBurnValueFromChangeMaps       
    :: (AssetId, TokenQuantity) -> NonEmpty TokenMap -> NonEmpty TokenMap   
removeBurnValueFromChangeMaps (assetId, assetQty) maps = maps       
    & fmap (`TokenMap.getQuantity` assetId) 
    & reduceTokenQuantities assetQty
    & NE.zipWith (`TokenMap.setQuantity` assetId) maps      
    
reduceTokenQuantities       
    :: TokenQuantity -> NonEmpty TokenQuantity -> NonEmpty TokenQuantity    
reduceTokenQuantities toBurn quantities =   
    NE.fromList $ burn toBurn (NE.toList quantities) []     
  where     
    burn _ [      ] ys = reverse ys 
    burn b (x : xs) ys      
        | x >= b = reverse ys <> (x' : xs)  
        | otherwise = burn b' xs (x' : ys)  
      where 
        b' = b `TokenQuantity.difference` x
        x' = x `TokenQuantity.difference` b    
```

Reply:
Addressed in af3622d43 / 89f100c40

Reply:
Addressed follow-up comment in c6b3146